scala/scala-api/src/org/jetbrains/plugins/scala/lang/psi/api/NewTypeInject.scala

```diff
@@ -0,0 +1,68 @@
+package org.jetbrains.plugins.scala.lang.psi.impl
+package toplevel
+package typedef
+
+import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.{ScClass, ScObject, ScTypeDefinition}
+import org.jetbrains.plugins.scala.lang.psi.types.{ScParameterizedType, ScType}
+
+class NewTypeInject extends SyntheticMembersInjector {
+
+  import NewTypeInject._
+
+  override def injectFunctions(source: ScTypeDefinition): Seq[String] = {
+    val companionClass = source match {
+      case obj: ScObject => obj.fakeCompanionClassOrCompanionClass
+      case _ => null
+    }
+
+    companionClass match {
```

Comment:
Annotations `@companionClass` control where an annotation on the implicit class will go. By default, annotations on an implicit class end up only on the class.